### PR TITLE
Refactor principal CLI commands

### LINKED ISSUE #59: Refactor CLI to more closely reflect Web interface
The current CLI consists of the following principal commands (there are more utility commands):

 * `build` - calculate, validate, and sign a partial GOBL Envelope
 * `envelope` - wrap in an envelope, calculate, validate, and sign a GOBL Document
 * `verify` - validate an envelope and check it was signed with a known signature.

The [Web UI](https://build.gobl.org) however is highlighting a different way to interact with GOBL envelopes and documents which feels a bit more intuitive. The following set of key actions will align both interfaces. All commands output a complete GOBL Envelope, regardless of the input which may be a partial envelope, or a document:

 * `calculate` - Perform calculations only, including invoice totals and envelope header digest.
 * `validate` - Check the contents of the envelope and document look correct (signatures are now optional in GOBL v0.28)
 * `sign` - requires that the envelope is no longer a draft, will calculate, validate and then sign the envelope.

With these set of changes, we can then focus on refining the meaning of the other commands:

 * `build` - performs a calculate and validate, but will not add a signature and works with draft documents.
 * `envelop` - no longer adds any meaning and can be removed.
 * `verify` - should continues as currently, but provide more explicit details on document validation and validity of each of the signatures.

### PR DESCRIPTION
**Pull request by dstotijn**
Closes #59. See that issue for details on what this PR encompasses.

## Tasks
- [x] Refactor `Build` command
- [x] Fix existing tests
- [x] Remove `Envelop` command
- [x] Add `internal.Sign` method
- [x] Write tests for `internal.Sign` command
- [x] Add `sign` CLI command
- [x] Write tests for `sign` CLI command
- [x] Add `internal.Calculate` method
- [x] Write tests for `internal.Calculate` command
- [x] Add `calculate` CLI command
- [x] Write tests for `calculate` CLI command

### CODE REVIEW

**Review by samlown — COMMENTED** [review]
Looking good in general! A few bits of duplicated code that could perhaps be simplified for the build opts. I think we've lost the "envelop" functionality though. We wanted to remove the command, but not the functionality which should now be supported by the "building" commands. Easiest to test from the console:

```bash
./gobl build ./samples/invoice-es.yaml
```
returns:
```
no-document
```

The expected result would be to automatically embed the document inside an envelope. The `gobl.Parse` method can help with this and will probably simplify quite a few bits of code.

File: internal/build.go [hunk 1]
```diff
@@ -45,66 +44,62 @@ func decodeInto(ctx context.Context, dest *map[string]interface{}, in io.Reader)
...
 	}
 
-	if err := finalizeEnvelope(ctx, env, opts); err != nil {
-		return nil, err
-	}
 	return env, nil
 }
 
-type schemaDataCB func(schema.ID) map[string]interface{}
+func parseBuildData(ctx context.Context, opts *BuildOptions) (*gobl.Envelope, error) {
```
Comment (dstotijn) [inline internal/build.go:72]: If we decide to rename `BuildOptions` to something that makes sense for the commands that need to parse incoming data, we can rename this func as well.

File: internal/sign.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+	"github.com/invopop/gobl"
+	"github.com/labstack/echo/v4"
+)
+
+// Sign parses a GOBL document into an envelope, performs calculations,
+// validates it, and finally signs its headers. The parsed envelope *must* be a
+// draft, or else an error is returned.
+func Sign(ctx context.Context, opts *BuildOptions) (*gobl.Envelope, error) {
+	// TODO: `BuildOptions` should probably be renamed to `ParseOptions`,
```
Comment (dstotijn) [inline internal/sign.go:15]: Is this `TODO` valid, and if so, do we want to include it in this PR?

**Comment by dstotijn** [comment]
> Looking good in general! A few bits of duplicated code that could perhaps be simplified for the build opts. I think we've lost the "envelop" functionality though. We wanted to remove the command, but not the functionality which should now be supported by the "building" commands. Easiest to test from the console:
> 
> ```shell
> ./gobl build ./samples/invoice-es.yaml
> ```
> 
> returns:
> 
> ```
> no-document
> ```
> 
> The expected result would be to automatically embed the document inside an envelope. The `gobl.Parse` method can help with this and will probably simplify quite a few bits of code.

This is now added to this PR. As discussed on Slack, unless `--type` is used, a root level `$schema` prop is now required and enforced during the parsing stage of `build`, `calculate` and `sign`.

File: internal/sign.go (hunk 2 above)
Comment (samlown) [reply internal/sign.go:15]: I think this is related to your previous comment and renaming "BuildOptions"... they don't really reflect what is happening.

**Review by samlown — APPROVED** [review]
Cool! This PR is a bit of a monster, so quite hard to verify completely.

Testing on the command line, I think the `sign` command should automatically remove the `draft` header, otherwise, you have to convert a file in two steps which is a bit annoying. For example, the current output is:

```bash
./gobl sign ./samples/invoice-es.yaml -i
cannot sign draft envelope
```

I'd expect a completely signed and finalized envelope.

I'm also slightly worried about the code duplication in each of the commands. My proposal would be to have a single overarching build process which is extended according to the option passed in.

**Comment by dstotijn** [comment]
> Testing on the command line, I think the `sign` command should automatically remove the `draft` header, otherwise, you have to convert a file in two steps which is a bit annoying. (...) I'd expect a completely signed and finalized envelope.

Ah, good catch! That would be nice. Pushed https://github.com/invopop/gobl.cli/pull/60/commits/1f4cd7f2b8a733329e06be39a651c6b2458275b6 which adds this behaviour.

> I'm also slightly worried about the code duplication in each of the commands. My proposal would be to have a single overarching build process which is extended according to the option passed in.

Yes, I agree. For now, I'll merge this PR to unblock other work if that's OK with you, and  I'll make sure we discuss how to consolidate/dedupe the logic next time we sync.

**Comment by samlown** [comment]
Sounds good. Let's get it merged 👍

**Pull request merged**